tensorflow_model_optimization/g3doc/guide/clustering/clustering_with_keras.ipynb

```diff
@@ -0,0 +1,1164 @@
+{
```

Comment:
<p>As mentioned in the email, the biggest change will be the restructuring of content according to the new guidelines.</p><p> </p><p>To avoid additional work, we may want to discuss top-level headers here first. Note that these documents will eventually be reviewed by others responsible for tensorflow.org docs in general, but I can give suggestions based on what was agreed for quantization-aware training.</p><p> </p><p>--------</p><p>Overview</p><ul><li>Other Pages</li><li>Contents</li></ul><p>Setup</p><p>Train a MNIST model without clustering</p><p>Clone and fine-tune pre-trained model with clustering</p><ul><li>Define the model</li><li>Train and evaluate the model against the baseline</li></ul><p>Create a compressed backend-independent model. (this is where you make the call to <code>strip_clustering</code>)</p><p>Export model to TFLite backend</p><p>See Nx smaller model from clustering</p><p>See Mx smaller model by combining post-training quantization with clustering</p><p>See persistence of accuracy from TF to TFLite</p><p>-----------</p><p> </p><p> </p><p>With the following considerations:</p><p>1) We could choose to see the persistence of accuracy twice, once for clustering by itself and once for clustering + quantization. For simplicity purposes, we do it just once after quantization.</p><p>2) Note that the headers are much more powerful (see Nx and see Mx smaller) and emphasize the story</p><p>as opposed to the actual implementation and code used.</p><p>3) From the <a href="https://github.com/tensorflow/docs/blob/master/tools/templates/notebook.ipynb" target="_blank">TF style guide</a>, note that H4 headers and below are not visible in the navigation.</p><p> </p><p>With regards to post-training quantization with clustering, I guess this is where we'll see that M is not 4x N given the per-channel aspects.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/model-optimization/pull/330/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Reply:
<p>Thanks.</p><p>All the bullet points are updated but there are two minor changes:</p><p>'Train and evaluate the model against the baseline' ---&gt; Cluster a whole model and evaluate the accuracy against the baseline</p><p>'See Nx smaller model from clustering' ---&gt; See 8x smaller model from clustering after compression. I think we only want to show size reduction after compression for the clustering case, right?</p><p> </p><p>'With regards to post-training quantization with clustering, I guess this is where we'll see that M is not 4x N given the per-channel aspects.' ---&gt; Which type of post-training quantization shoud be used here? Does it have to be the same with the one in the QAT or the pruning example?</p>

Reply:
>  See 8x smaller model from clustering after compression. I think we only want to show size reduction after compression for the clustering case, right?

Yes after applying zip. "after compression" is confusing by itself since you could think that clustering is doing the compression.

> Which type of post-training quantization shoud be used here?
Anything is fine. Both dynamic range and integer post-training quantization in TFLite are using per-axis quantization going forward, so you shouldn't use that to make a decision.

Reply:
<p>Thanks, will use 'after applying zip' and any post-training quantization.</p>